### PR TITLE
mon: --format=json does not work for config get or show

### PR DESCRIPTION
**Pull request by zhengchengyao**
Fixes: https://tracker.ceph.com/issues/44089
Signed-off-by: Zheng Yin <zhengyin@cmss.chinamobile.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by zhengchengyao** [comment]
jenkins test docs

**Comment by zhengchengyao** [comment]
jenkins test signed

**Comment by zhengchengyao** [comment]
jenkins test submodules

File: src/mgr/DaemonServer.cc [hunk 1]
```diff
@@ -1812,12 +1827,12 @@ bool DaemonServer::_handle_command(
...
       if (p != daemon->config.end() &&
 	  !p->second.empty()) {
-	cmdctx->odata.append(p->second.rbegin()->second + "\n");
+	print_value(f.get(), cmdctx->odata, name, p->second.rbegin()->second);
       } else {
 	auto& defaults = daemon->_get_config_defaults();
 	auto q = defaults.find(name);
 	if (q != defaults.end()) {
-	  cmdctx->odata.append(q->second + "\n");
+	  print_value(f.get(), cmdctx->odata, name, q->second);
```
Comment (batrick) [inline src/mgr/DaemonServer.cc]: `print_value` should be reused for the vanilla `config show <entity>`  as well (below this code).

File: src/mgr/DaemonServer.cc [hunk 2]
```diff
@@ -837,6 +837,21 @@ void DaemonServer::log_access_denied(
         "#client-authentication";
 }
 
+static void print_value(Formatter *f, bufferlist &odata,
+                        const string &name, const string &value)
+{
+  if (f) {
+    f->open_object_section("config");
+    f->dump_string("name", name);
+    f->dump_string("value", value);
```
Comment (batrick) [inline src/mgr/DaemonServer.cc]: include `source`, `overrides` as below in vanilla `config show <entity>`. I think `default `would also be valuable.

File: src/mon/ConfigMonitor.cc [hunk 3]
```diff
@@ -141,6 +141,21 @@ static string indent_who(const string& who)
   return "    " + who;
 }
 
+static void print_value(Formatter *f, bufferlist &odata,
+                        const string &name, const string &value)
+{
+  if (f) {
+    f->open_object_section("config");
+    f->dump_string("name", name);
+    f->dump_string("value", value);
```
Comment (batrick) [inline src/mon/ConfigMonitor.cc]: as I noted for `config show`, we should reuse `print_value` for the vanilla `config get <entity>` and include any fields it has missing here.

File: src/mgr/DaemonServer.cc (hunk 1 above)
Comment (zhengchengyao) [reply src/mgr/DaemonServer.cc]: thanks， I will reused print_value.

File: src/mon/ConfigMap.cc [hunk 4]
```diff
@@ -204,6 +204,20 @@ bool ConfigMap::parse_mask(
...
 
+void ConfigMap::print_value(Formatter *f, bufferlist &odata,
+                 const string &name, const string &value)
+{
+  if (f) {
+    f->open_object_section("config");
+    f->dump_string("name", name);
+    f->dump_string("value", value);
+    f->close_section();
+    f->flush(odata);
```
Comment (batrick) [inline src/mon/ConfigMap.cc]: Please also include the information from:

https://github.com/ceph/ceph/blob/d6a36b014bf1f447801ad009d0677022c61645c9/src/mon/ConfigMonitor.cc#L356-L362

and update that code to also call `print_value`.

Comment (zhengchengyao) [reply src/mon/ConfigMap.cc]: print_value includes mask,section,can_update_at_runtime, in mgr/DaemonServer.cc, it also includes  source, overrides. 
I don't know how to unify. Do you have any good ideas？

**Comment by neha-ojha** [comment]
@zhengchengyao would like to revisit this PR and address code review comments?

**Comment by guits** [comment]
@neha-ojha @batrick this was closed because of no activity, any chance someone can take over this patch?
Thanks!

**Comment by rzarzynski** [comment]
CC: @leonid-s-usov.

**Comment by leonid-s-usov** [comment]
Yep, I'm assigned this issue now, will pick this up

**Pull request closed without merging**